spec/services/submission/trait_score_parser_spec.rb

```diff
@@ -5,101 +5,234 @@
   let(:upload) { create(:upload) }
   subject { described_class.new(upload) }
 
-  describe '#map_headers_to_traits' do
+  describe '#parse_header' do
     context 'provided with input of incomplete content' do
-      it 'ignores single column header' do
-        input_is 'single column name no tabs'
-        subject.send(:map_headers_to_traits)
+      it 'reports less than three-column header as incomplete' do
+        input_is 'single column name'
+        subject.send(:parse_header)
+        expect(upload.errors[:file]).
+          to eq ['No correct header provided. At least four columns are expected.']
+      end
+
+      it 'does nothing with a three-column header' do
+        input_is 'single column name,Plant accession,oo,pl'
+        subject.send(:parse_header)
         expect(subject.trait_mapping).to eq({})
       end
 
-      it 'ignores all columns when no traits chosen' do
-        input_is "id,first trait,\"second,trait_name\""
-        subject.send(:map_headers_to_traits)
+      it 'ignores all columns when no traits were chosen' do
+        input_is "id,Plant accession,oo,pl,first trait,\"second,trait_name\""
+        subject.send(:parse_header)
         expect(subject.trait_mapping).to eq({})
       end
+
+      it 'complains if there is no Plant accession column' do
+        input_is 'id,pa,oo,pl'
+        subject.send(:parse_header)
+        expect(upload.errors[:file]).
+          to eq ['No correct header provided. Please provide the \"Plant accession\" column.']
+      end
+
+      it 'complains if there is neither Plant line nor Plant variety column' do
+        input_is 'id,Plant accession,oo,pl'
+        subject.send(:parse_header)
+        expect(upload.errors[:file]).
+          to eq ['No correct header provided. Please provide either the \"Plant line\" or the \"Plant variety\" column.']
+      end
     end
 
     context 'provided with trait-rich submission' do
       it 'maps columns by name' do
         upload.submission.content.update(:step02, trait_descriptor_list: ['trait 1', 'trait 2'])
-        input_is "id,trait 2,trait 1"
-        subject.send(:map_headers_to_traits)
+        input_is "id,Plant accession,oo,Plant line,trait 2,trait 1"
+        subject.send(:parse_header)
         expect(subject.trait_mapping).
           to eq({0 => 1, 1 => 0})
       end
 
       it 'recognizes names with commas and surrounding white chars' do
         upload.submission.content.update(:step02, trait_descriptor_list: ['trait,1', 'trait 2'])
-        input_is "id,trait 2  ,\"trait,1\""
-        subject.send(:map_headers_to_traits)
+        input_is "id,Plant accession,oo,Plant line,trait 2  ,\"trait,1\""
+        subject.send(:parse_header)
         expect(subject.trait_mapping).
           to eq({0 => 1, 1 => 0})
       end
 
       it 'honors proper trait sorting by index' do
         upload.submission.content.update(:step02, trait_descriptor_list: ['Ctrait', 'Atrait', 'Btrait'])
```

Comment:
Yes, this maps column ordinal numbers (starting with 0 for the first scoring column). There should be two cases supported:
 - mapping in any order as long as the user keeps the exact trait names
 - mapping in exact trait order if the user does not do that (and you know - the encoding hell... ;)).

Tech reps, however, are expected in natural order anyway (I don't think anyone would start with rep2 in order to follow with rep1 for the same trait...).